File: src/core/hough.ts
// F-01 直線検出器(Hough 変換)。客観層。
//
// 座標系は画像中心原点、x cosθ + y sinθ = ρ、θ ∈ [0, π)。
// **(θ, ρ) と (θ+π, −ρ) は同じ直線である。**比較は lineDistance で畳んでから行う。
//
// 真値を参照しない(G-03 / T-007)。しきい値は引数であり、既定値は DEFAULT_DETECT
// 一箇所にだけ置く —— ① のスライダーはここを動かす。

import { canny, DEFAULT_CANNY, type CannyOptions } from "./canny";
import type { Line, RasterImage } from "./image";

export type AccumulateCtx = {
  cx: number;
  cy: number;
  thetaSteps: number;
  rhoSteps: number;
  /** ρ 1 px あたりのビン数 */
  rhoScale: number;
  cosT: Float64Array;
  sinT: Float64Array;
};

export type AccumulateFn = (acc: Int32Array, ctx: AccumulateCtx, x: number, y: number) => void;

export const defaultAccumulate: AccumulateFn = (acc, ctx, x, y) => {
  // 画素中心を使う(image.ts の toCenter と同じ規約)
  const dx = x + 0.5 - ctx.cx;
  const dy = y + 0.5 - ctx.cy;
  const half = ctx.rhoSteps >> 1;
  for (let t = 0; t < ctx.thetaSteps; t++) {
    const rho = dx * ctx.cosT[t] + dy * ctx.sinT[t];
    const r = Math.round(rho * ctx.rhoScale) + half;
    if (r >= 0 && r < ctx.rhoSteps) acc[t * ctx.rhoSteps + r]++;
  }
};

export type DetectOptions = {
  canny: CannyOptions;
  /** θ のビン数。180 なら 1° 刻み */
  thetaSteps: number;
  /** ρ 1 px あたりのビン数 */
  rhoScale: number;
  /** 最大票数に対する比。① のしきい値スライダーはここを動かす */
  voteRatio: number;
  /** 非極大抑制の窓(θ ビン / ρ ビン) */
  nmsTheta: number;
  nmsRho: number;
  /** 返す直線の上限 */
  maxLines: number;
  /** 差し替え可能な投票関数。陽性対照(壊した版)を差し込む口(T-004) */
  accumulate?: AccumulateFn;
  /**
   * 角度表を外から与える。**二実装照合で使う**(G-06a)。
   *
   * cos/sin は Rust と V8 で最終 ULP が食い違う(実測 2026-08-31: 44/2000・50/2000)。
   * 自前で計算したままだと「アルゴリズムの差」と「libm の差」が混ざるので、
   * 照合のときだけ同じ表を両実装へ渡す。実運用ではこの口を使わない。
   */
  angleTable?: { cos: Float64Array; sin: Float64Array };
};

// 既定値は実測(2026-08-31)で決めた。
//
// **分解能**: thetaSteps=180(1° ビン)/ rhoScale=1(1 px ビン)。900 ケース掃引で
// 最悪 Δθ=0.137° / Δρ=0.537 px。360/2 に上げると Δρ=0.801 px と **悪くなる** ——
// ビンを細かくすると票が分散し、±1 ビンの重心補間が拾う裾が痩せるためである。
// 「分解能を上げれば精度が上がる」は、重心補間を挟むと成り立たない。
//
// **voteRatio**: 0.25。5 本同時の検査(T-003)で、0.5 では弦の短い線が下限を割って
// 5 本中 3 本しか復元できなかった。0.25 で 5 本すべてを復元し、誤検出は 0 本。
// 0.1 まで下げると 7 本になり誤検出が出る。① のしきい値スライダーはここを動かす ——
// **「その絵の構図線」が分析者の選択の関数であること**を、この値が体現している。
export const DEFAULT_DETECT: DetectOptions = {
  canny: DEFAULT_CANNY,
  thetaSteps: 180,
  rhoScale: 1,
  voteRatio: 0.25,
  nmsTheta: 3,
  nmsRho: 4,
  maxLines: 32,
};

export type HoughResult = {
  acc: Int32Array;
  ctx: AccumulateCtx;
};

export function houghTransform(
  edges: Uint8Array,
  width: number,
  height: number,
  opts: DetectOptions,
): HoughResult {
  const cx = width / 2;
  const cy = height / 2;
  const R = Math.hypot(cx, cy);
  const rhoSteps = 2 * Math.ceil(R * opts.rhoScale) + 1;
  let cosT: Float64Array;
  let sinT: Float64Array;
  if (opts.angleTable) {
    if (opts.angleTable.cos.length !== opts.thetaSteps || opts.angleTable.sin.length !== opts.thetaSteps) {
      throw new Error(
        `角度表の長さ(${opts.angleTable.cos.length}/${opts.angleTable.sin.length})が thetaSteps=${opts.thetaSteps} と合わない`,
      );
    }
    cosT = opts.angleTable.cos;
    sinT = opts.angleTable.sin;
  } else {
    cosT = new Float64Array(opts.thetaSteps);
    sinT = new Float64Array(opts.thetaSteps);
    for (let t = 0; t < opts.thetaSteps; t++) {
      const th = (t * Math.PI) / opts.thetaSteps;
      cosT[t] = Math.cos(th);
      sinT[t] = Math.sin(th);
    }
  }
  const ctx: AccumulateCtx = {
    cx,
    cy,
    thetaSteps: opts.thetaSteps,
    rhoSteps,
    rhoScale: opts.rhoScale,
    cosT,
    sinT,
  };
  const acc = new Int32Array(opts.thetaSteps * rhoSteps);
  const accumulate = opts.accumulate ?? defaultAccumulate;
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      if (edges[y * width + x]) accumulate(acc, ctx, x, y);
    }
  }
  return { acc, ctx };
}

/**
 * 極大抽出。ビンの重心で補間して量子化誤差を落とす。
 * θ は端で巻き込む —— t = 0 の隣は t = thetaSteps − 1 で、そこでは ρ の符号が反転する。
 */
export function extractPeaks(res: HoughResult, opts: DetectOptions): Line[] {
  const { acc, ctx } = res;
  const { thetaSteps, rhoSteps } = ctx;
  let max = 0;
  for (let i = 0; i < acc.length; i++) if (acc[i] > max) max = acc[i];
  if (max === 0) return [];
  const minVotes = Math.max(1, max * opts.voteRatio);

  const cand: { t: number; r: number; v: number }[] = [];
  for (let t = 0; t < thetaSteps; t++) {
    for (let r = 1; r < rhoSteps - 1; r++) {
      const v = acc[t * rhoSteps + r];
      if (v < minVotes) continue;
      // 局所最大か(θ は巻き込み、ρ は境界を除いてある)
      let isMax = true;
      for (let dt = -1; dt <= 1 && isMax; dt++) {
        for (let dr = -1; dr <= 1; dr++) {
          if (dt === 0 && dr === 0) continue;
          const [tt, rr] = wrap(t + dt, r + dr, thetaSteps, rhoSteps);
          if (tt < 0) continue;
          if (acc[tt * rhoSteps + rr] > v) {
            isMax = false;
            break;
          }
        }
      }
      if (isMax) cand.push({ t, r, v });
    }
  }
  cand.sort((a, b) => b.v - a.v || a.t - b.t || a.r - b.r);

  const kept: { t: number; r: number; v: number }[] = [];
  for (const c of cand) {
    if (kept.length >= opts.maxLines) break;
    const near = kept.some((k) => {
      const dt = circularBinDistance(k.t, c.t, thetaSteps);
      // θ が巻き込んだ組では ρ の符号が反転している
      const flipped = Math.abs(k.t - c.t) > thetaSteps / 2;
      const kr = flipped ? rhoSteps - 1 - k.r : k.r;
      return dt <= opts.nmsTheta && Math.abs(kr - c.r) <= opts.nmsRho;
    });
    if (!near) kept.push(c);
  }

  return kept.map((k) => refine(acc, ctx, k.t, k.r));
}

function wrap(t: number, r: number, thetaSteps: number, rhoSteps: number): [number, number] {
  if (r < 0 || r >= rhoSteps) return [-1, -1];
  if (t >= 0 && t < thetaSteps) return [t, r];
  // θ が範囲外 —— 反対端へ巻き込み、ρ の符号を反転する
  const tt = ((t % thetaSteps) + thetaSteps) % thetaSteps;
  const rr = rhoSteps - 1 - r;
  return [tt, rr];
}

function circularBinDistance(a: number, b: number, n: number): number {
  const d = Math.abs(a - b);
  return Math.min(d, n - d);
}

/** ±1 ビンの票で重心を取り、ビンの中心からのずれを補う */
function refine(acc: Int32Array, ctx: AccumulateCtx, t: number, r: number): Line {
  const { thetaSteps, rhoSteps, rhoScale } = ctx;
  let wt = 0;
  let wr = 0;
  let sum = 0;
  for (let dt = -1; dt <= 1; dt++) {
    for (let dr = -1; dr <= 1; dr++) {
      const [tt, rr] = wrap(t + dt, r + dr, thetaSteps, rhoSteps);
      if (tt < 0) continue;
      const v = acc[tt * rhoSteps + rr];
      // 巻き込んだ側は ρ 軸が反転しているので、重心の寄与も反転させる
      const flipped = t + dt < 0 || t + dt >= thetaSteps;
      wt += v * dt;
      wr += v * (flipped ? -dr : dr);
      sum += v;
    }
  }
  const ft = sum > 0 ? t + wt / sum : t;
  const fr = sum > 0 ? r + wr / sum : r;
  const theta = (ft * Math.PI) / thetaSteps;
  const rho = (fr - (rhoSteps >> 1)) / rhoScale;
  return normalizeLine({ theta, rho });
}

/** θ を [0, π) に畳む。畳むとき ρ の符号が反転する */
export function normalizeLine(l: Line): Line {
  let theta = l.theta;
  let rho = l.rho;
  const PI = Math.PI;
  while (theta < 0) {
    theta += PI;
    rho = -rho;
  }
  while (theta >= PI) {
    theta -= PI;
    rho = -rho;
  }
  return { theta, rho };
}

export type LineDistance = {
  dThetaDeg: number;
  dRho: number;
  /** θ を px に換算して足した合成距離。最近傍を選ぶためだけに使う */
  combined: number;
};

/** (θ+π, −ρ) の同値を畳んでから 2 直線の隔たりを測る */
export function lineDistance(a: Line, b: Line, radius: number): LineDistance {
  const A = normalizeLine(a);
  const B = normalizeLine(b);
  const raw = Math.abs(A.theta - B.theta);
  const wrapped = Math.PI - raw;
  let dTheta: number;
  let dRho: number;
  if (raw <= wrapped) {
    dTheta = raw;
    dRho = Math.abs(A.rho - B.rho);
  } else {
    dTheta = wrapped;
    dRho = Math.abs(A.rho + B.rho);
  }
  const dThetaDeg = (dTheta * 180) / Math.PI;
  return { dThetaDeg, dRho, combined: dRho + dTheta * radius };
}

export function detectLines(img: RasterImage, opts: DetectOptions = DEFAULT_DETECT): Line[] {
  const edges = canny(img, opts.canny);
  const res = houghTransform(edges, img.width, img.height, opts);
  return extractPeaks(res, opts);
}

File: rust/src/hough.rs
//! Hough 投票と極大抽出の Rust 実装。TS 実装(src/core/hough.ts)と突き合わせる。
//!
//! **この区間は整数と四則だけで書く。**三角関数は呼ばず、角度表を引数で受け取る ——
//! 実測(2026-08-31)で cos/sin は Rust と V8 で最終 ULP が食い違うため、
//! 自前で計算すると「アルゴリズムの差」と「libm の差」が分離できなくなる(SPEC §11.1)。

/// 投票平面。`acc[t * rho_steps + r]`
pub struct Hough {
    pub acc: Vec<i32>,
    pub theta_steps: usize,
    pub rho_steps: usize,
    pub rho_scale: f64,
}

/// エッジ座標(画像左上原点の整数)から投票平面を作る。
/// 原点は画像中心。`cos_t` / `sin_t` は呼び出し側が与える。
pub fn accumulate(
    edges: &[(u32, u32)],
    width: usize,
    height: usize,
    cos_t: &[f64],
    sin_t: &[f64],
    rho_scale: f64,
) -> Hough {
    let theta_steps = cos_t.len();
    assert_eq!(theta_steps, sin_t.len(), "角度表の長さが揃っていない");
    let cx = width as f64 / 2.0;
    let cy = height as f64 / 2.0;
    let r = (cx * cx + cy * cy).sqrt();
    let rho_steps = 2 * (r * rho_scale).ceil() as usize + 1;
    let half = (rho_steps >> 1) as i64;
    let mut acc = vec![0i32; theta_steps * rho_steps];
    for &(x, y) in edges {
        // 画素中心を使う(TS の image.ts::toCenter と同じ規約)
        let dx = x as f64 + 0.5 - cx;
        let dy = y as f64 + 0.5 - cy;
        for t in 0..theta_steps {
            let rho = dx * cos_t[t] + dy * sin_t[t];
            let r_bin = round_half_away(rho * rho_scale) + half;
            if r_bin >= 0 && (r_bin as usize) < rho_steps {
                acc[t * rho_steps + r_bin as usize] += 1;
            }
        }
    }
    Hough {
        acc,
        theta_steps,
        rho_steps,
        rho_scale,
    }
}

/// JavaScript の `Math.round` は **半整数を +∞ 方向へ丸める**(-0.5 → -0)。
/// Rust の `f64::round` は 0 から遠い方へ丸める(-0.5 → -1)。**負値で食い違う。**
/// TS 実装に合わせる。
pub fn round_half_away(v: f64) -> i64 {
    (v + 0.5).floor() as i64
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Line {
    pub theta: f64,
    pub rho: f64,
}

/// θ を [0, π) に畳む。畳むとき ρ の符号が反転する
pub fn normalize_line(mut theta: f64, mut rho: f64) -> Line {
    let pi = std::f64::consts::PI;
    while theta < 0.0 {
        theta += pi;
        rho = -rho;
    }
    while theta >= pi {
        theta -= pi;
        rho = -rho;
    }
    Line { theta, rho }
}

fn wrap(t: i64, r: i64, theta_steps: usize, rho_steps: usize) -> Option<(usize, usize)> {
    if r < 0 || r >= rho_steps as i64 {
        return None;
    }
    if t >= 0 && t < theta_steps as i64 {
        return Some((t as usize, r as usize));
    }
    let ts = theta_steps as i64;
    let tt = ((t % ts) + ts) % ts;
    let rr = rho_steps as i64 - 1 - r;
    Some((tt as usize, rr as usize))
}

fn circular_bin_distance(a: usize, b: usize, n: usize) -> usize {
    let d = if a > b { a - b } else { b - a };
    d.min(n - d)
}

pub struct PeakOptions {
    pub vote_ratio: f64,
    pub nms_theta: usize,
    pub nms_rho: usize,
    pub max_lines: usize,
}

/// 極大抽出。**PI の乗除しか使わないので TS と完全一致を要求できる**
pub fn extract_peaks(h: &Hough, opts: &PeakOptions) -> Vec<Line> {
    let (ts, rs) = (h.theta_steps, h.rho_steps);
    let max = h.acc.iter().copied().max().unwrap_or(0);
    if max == 0 {
        return Vec::new();
    }
    let min_votes = (max as f64 * opts.vote_ratio).max(1.0);

    let mut cand: Vec<(usize, usize, i32)> = Vec::new();
    for t in 0..ts {
        for r in 1..rs - 1 {
            let v = h.acc[t * rs + r];
            if (v as f64) < min_votes {
                continue;
            }
            let mut is_max = true;
            'outer: for dt in -1i64..=1 {
                for dr in -1i64..=1 {
                    if dt == 0 && dr == 0 {
                        continue;
                    }
                    if let Some((tt, rr)) = wrap(t as i64 + dt, r as i64 + dr, ts, rs) {
                        if h.acc[tt * rs + rr] > v {
                            is_max = false;
                            break 'outer;
                        }
                    }
                }
            }
            if is_max {
                cand.push((t, r, v));
            }
        }
    }
    cand.sort_by(|a, b| b.2.cmp(&a.2).then(a.0.cmp(&b.0)).then(a.1.cmp(&b.1)));

    let mut kept: Vec<(usize, usize, i32)> = Vec::new();
    for c in cand {
        if kept.len() >= opts.max_lines {
            break;
        }
        let near = kept.iter().any(|k| {
            let dt = circular_bin_distance(k.0, c.0, ts);
            let flipped = (k.0 as i64 - c.0 as i64).abs() > (ts / 2) as i64;
            let kr = if flipped { rs - 1 - k.1 } else { k.1 };
            dt <= opts.nms_theta && (kr as i64 - c.1 as i64).abs() <= opts.nms_rho as i64
        });
        if !near {
            kept.push(c);
        }
    }
    kept.into_iter().map(|k| refine(h, k.0, k.1)).collect()
}

fn refine(h: &Hough, t: usize, r: usize) -> Line {
    let (ts, rs) = (h.theta_steps, h.rho_steps);
    let mut wt = 0.0f64;
    let mut wr = 0.0f64;
    let mut sum = 0.0f64;
    for dt in -1i64..=1 {
        for dr in -1i64..=1 {
            if let Some((tt, rr)) = wrap(t as i64 + dt, r as i64 + dr, ts, rs) {
                let v = h.acc[tt * rs + rr] as f64;
                let flipped = t as i64 + dt < 0 || t as i64 + dt >= ts as i64;
                wt += v * dt as f64;
                wr += v * if flipped { -(dr as f64) } else { dr as f64 };
                sum += v;
            }
        }
    }
    let ft = if sum > 0.0 {
        t as f64 + wt / sum
    } else {
        t as f64
    };
    let fr = if sum > 0.0 {
        r as f64 + wr / sum
    } else {
        r as f64
    };
    let theta = ft * std::f64::consts::PI / ts as f64;
    let rho = (fr - (rs >> 1) as f64) / h.rho_scale;
    normalize_line(theta, rho)
}

#[cfg(test)]
mod tests {
    use super::*;

    // 期待値の出所: JS の `Math.round` の仕様(ECMA-262 —— 半整数は +∞ 方向へ丸める)。
    // Rust の `f64::round` は 0 から遠い方へ丸めるので **負の半整数で食い違う**。
    #[test]
    fn round_matches_javascript_math_round() {
        assert_eq!(round_half_away(0.5), 1);
        assert_eq!(round_half_away(1.5), 2);
        assert_eq!(round_half_away(-0.5), 0); // JS: Math.round(-0.5) === -0
        assert_eq!(round_half_away(-1.5), -1); // JS: Math.round(-1.5) === -1
        assert_eq!(round_half_away(-2.5), -2);
        // 陽性対照: f64::round はここで違う値を出す(検査が働いていることの確認)
        assert_ne!((-0.5f64).round() as i64, round_half_away(-0.5));
        assert_ne!((-1.5f64).round() as i64, round_half_away(-1.5));
    }

    #[test]
    fn normalize_folds_theta_into_half_turn() {
        let pi = std::f64::consts::PI;
        let a = normalize_line(pi + 0.3, 5.0);
        assert!((a.theta - 0.3).abs() < 1e-12);
        assert!((a.rho + 5.0).abs() < 1e-12);
        let b = normalize_line(-0.2, 5.0);
        assert!((b.theta - (pi - 0.2)).abs() < 1e-12);
        assert!((b.rho + 5.0).abs() < 1e-12);
    }

    #[test]
    fn uniform_input_yields_no_peaks() {
        // エッジが 1 つも無ければ極大も無い(分母 0 の縮退 — HC-097)
        let h = accumulate(&[], 32, 32, &[1.0], &[0.0], 1.0);
        let p = extract_peaks(
            &h,
            &PeakOptions {
                vote_ratio: 0.25,
                nms_theta: 3,
                nms_rho: 4,
                max_lines: 8,
            },
        );
        assert!(p.is_empty());
    }
}
